projects/igniteui-angular/src/lib/grids/filtering/grid-filtering-cell.component.ts

```diff
@@ -0,0 +1,311 @@
+import {
+    ChangeDetectorRef,
+    Component,
+    Input,
+    TemplateRef,
+    ViewChild,
+    HostBinding,
+    AfterViewInit,
+    ElementRef,
+    HostListener,
+    OnInit
+} from '@angular/core';
+import { IgxColumnComponent } from '../column.component';
+import { IFilteringExpression } from '../../data-operations/filtering-expression.interface';
+import { FilteringExpressionsTree } from '../../data-operations/filtering-expressions-tree';
+import { IBaseChipEventArgs, IgxChipsAreaComponent, IgxChipComponent } from '../../chips';
+import { IgxFilteringService, ExpressionUI } from './grid-filtering.service';
+import { KEYS, cloneArray } from '../../core/utils';
+import { IgxGridNavigationService } from '../grid-navigation.service';
+
+/**
+ * @hidden
+ */
+@Component({
+    preserveWhitespaces: false,
+    selector: 'igx-grid-filtering-cell',
+    templateUrl: './grid-filtering-cell.component.html'
+})
+export class IgxGridFilteringCellComponent implements AfterViewInit, OnInit {
+
+    private rootExpressionsTree: FilteringExpressionsTree;
+    private expressionsList: ExpressionUI[];
+    private baseClass = 'igx-grid__filtering-cell-indicator';
+    private currentTemplate = null;
+
+    public visibleExpressionsList: ExpressionUI[];
+    public moreFiltersCount = 0;
+
+    @Input()
+    public column: IgxColumnComponent;
+
+    @ViewChild('emptyFilter', { read: TemplateRef })
+    protected emptyFilter: TemplateRef<any>;
+
+    @ViewChild('defaultFilter', { read: TemplateRef })
+    protected defaultFilter: TemplateRef<any>;
+
+    @ViewChild('complexFilter', { read: TemplateRef })
+    protected complexFilter: TemplateRef<any>;
+
+    @ViewChild('chipsArea', { read: IgxChipsAreaComponent })
+    protected chipsArea: IgxChipsAreaComponent;
+
+    @ViewChild('moreIcon', { read: ElementRef })
+    protected moreIcon: ElementRef;
+
+    @ViewChild('ghostChip', { read: IgxChipComponent })
+    protected ghostChip: IgxChipComponent;
+
+    @ViewChild('complexChip', { read: IgxChipComponent })
+    protected complexChip: IgxChipComponent;
+
+    @HostBinding('style.min-width')
+    @HostBinding('style.max-width')
+    @HostBinding('style.flex-basis')
+    get width() {
+        // HACK - think of a better solution
+        const colWidth = this.column.width;
+        const isPercentageWidth = colWidth && typeof colWidth === 'string' && colWidth.indexOf('%') !== -1;
+
+        if (isPercentageWidth) {
+            const firstContentCell = this.column.cells[0];
+            if (firstContentCell) {
+                return firstContentCell.nativeElement.getBoundingClientRect().width + 'px';
+            }
+        } else {
+            return this.column.width;
+        }
+    }
+
+    @HostBinding('class.igx-grid__filtering-cell')
+    public cssClass = 'igx-grid__filtering-cell';
+
+    @HostBinding('class.igx-grid__th--pinned-last')
+    get isLastPinned() {
```

Comment:
This function is repeated across several components. Let's think of a way to have it in one place.